src/set.rs

```diff
@@ -0,0 +1,1114 @@
+        if cfg!(not(feature = "test_debug")) {
+            f.debug_set().entries(self.iter()).finish()
+        } else {
+            // Let the inner `OrderMap` print all of its details
+            f.debug_struct("OrderSet").field("map", &self.map).finish()
+        }
+    }
+}
+
+impl<T> OrderSet<T> {
+    /// Create a new set. (Does not allocate.)
+    pub fn new() -> Self {
+        OrderSet { map: OrderMap::new() }
+    }
+
+    /// Create a new set with capacity for `n` elements.
+    /// (Does not allocate if `n` is zero.)
+    ///
+    /// Computes in **O(n)** time.
+    pub fn with_capacity(n: usize) -> Self {
+        OrderSet { map: OrderMap::with_capacity(n) }
+    }
+}
+
+impl<T, S> OrderSet<T, S> {
+    /// Create a new set with capacity for `n` elements.
+    /// (Does not allocate if `n` is zero.)
+    ///
+    /// Computes in **O(n)** time.
+    pub fn with_capacity_and_hasher(n: usize, hash_builder: S) -> Self
+        where S: BuildHasher
+    {
+        OrderSet { map: OrderMap::with_capacity_and_hasher(n, hash_builder) }
+    }
+
+    /// Return the number of elements in the set.
+    ///
+    /// Computes in **O(1)** time.
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Returns true if the set contains no elements.
+    ///
+    /// Computes in **O(1)** time.
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    /// Create a new set with `hash_builder`
+    pub fn with_hasher(hash_builder: S) -> Self
+        where S: BuildHasher
+    {
+        OrderSet { map: OrderMap::with_hasher(hash_builder) }
+    }
+
+    /// Return a reference to the set's `BuildHasher`.
+    pub fn hasher(&self) -> &S
+        where S: BuildHasher
+    {
+        self.map.hasher()
+    }
+
+    /// Computes in **O(1)** time.
+    pub fn capacity(&self) -> usize {
+        self.map.capacity()
+    }
+}
+
+impl<T, S> OrderSet<T, S>
+    where T: Hash + Eq,
+          S: BuildHasher,
+{
+    /// Remove all elements in the set, while preserving its capacity.
+    ///
+    /// Computes in **O(n)** time.
+    pub fn clear(&mut self) {
+        self.map.clear();
+    }
+
+    /// FIXME Not implemented fully yet
+    pub fn reserve(&mut self, additional: usize) {
+        self.map.reserve(additional);
+    }
+
+    /// Insert the value into the map.
+    ///
+    /// If the value did already exist in the set, returns `false`;
+    /// otherwise, return `true`.
+    ///
+    /// Computes in **O(1)** time (amortized average).
+    pub fn insert(&mut self, value: T) -> bool {
+        self.map.insert(value, ()).is_none()
+    }
+
+    /// Return an iterator over the values of the set, in their order
+    pub fn iter(&self) -> Iter<T> {
+        Iter {
+            iter: self.map.keys()
+        }
+    }
+
+    /// Return an iterator over the values that are in `self` but not `other`.
+    pub fn difference<'a, S2>(&'a self, other: &'a OrderSet<T, S2>) -> Difference<'a, T, S2>
```

Comment:
We need a comment on the order, not sure if there's anything interesting we can say. If we prefer to not guarantee anything that's ok. Difference, in particular, seems like a simple example of guaranteeing the order -- same as self, whenever an element appears.

Reply:
I think it's fine to guarantee order on all of them.  The tests I wrote do make sure of this, since they compare the iterators directly against range iterators.

Reply:
I've documented them all.